lib/xnvme_be_linux_hugemem.c

```diff
@@ -0,0 +1,173 @@
+	char line[1024];
+	char search_str[1024];
+	FILE *fp;
+
+	fp = fopen("/proc/mounts", "r");
+
+	strncpy(search_str, path, sizeof(search_str) - 1);
+	strncat(search_str, " hugetlbfs", sizeof(search_str) - 1);
+
+	while (fgets(line, sizeof(line), fp)) {
+		if (strstr(line, search_str)) {
+			return 1;
+		}
+	}
+	return 0;
+}
+
+void *
+xnvme_be_linux_buf_alloc(const struct xnvme_dev *XNVME_UNUSED(dev), size_t nbytes,
+			 uint64_t *XNVME_UNUSED(phys))
+{
+	void *buf;
+	int hugepage_fd;
+	struct huge_alloc *entry;
+	char hugepage_path[1024];
+	size_t hugepage_size = get_hugepage_size();
+	char *env_hugepage_path = getenv("XNVME_HUGETLB_PATH");
+
+	if (env_hugepage_path == NULL) {
+		XNVME_DEBUG("ERROR: XNVME_HUGETLB_PATH env var not defined");
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	strncpy(hugepage_path, env_hugepage_path, sizeof(hugepage_path) - 1);
+
+	if (!verify_hugetlbfs_path(hugepage_path)) {
+		XNVME_DEBUG("WARNING: Hugetlbfs is not mounted at: %s", hugepage_path);
+	}
+
+	if (!hugepage_size) {
+		XNVME_DEBUG("Hugepage size not valid: %lu", hugepage_size);
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	// nbytes has to be a multiple of alignment. Therefore, we round up to the nearest
+	// multiple.
+	nbytes = (1 + ((nbytes - 1) / hugepage_size)) * (hugepage_size);
+
+	// Map a Huge page
+	strncat(hugepage_path, "/xnvme_XXXXXX", sizeof(hugepage_path) - 1);
+	hugepage_fd = mkstemp(hugepage_path);
+	XNVME_DEBUG("opened hugepage file %s", hugepage_path);
+	if (hugepage_fd == -1) {
+		XNVME_DEBUG("Failed to open hugepage file. %s", hugepage_path);
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	if (ftruncate(hugepage_fd, nbytes)) {
+		XNVME_DEBUG("Failed to truncate hugepage file. %s, %lu", hugepage_path, nbytes);
+		perror("hugetlb ftruncate()");
+		return NULL;
+	}
+
+	buf = mmap(NULL, nbytes, PROT_READ | PROT_WRITE, MAP_SHARED, hugepage_fd, 0);
+	if (buf == MAP_FAILED) {
+		XNVME_DEBUG("mmap failed on hugepage file");
+		return NULL;
+	}
+
+	// Appending allocated hugepage to linked-list for use during free
+	entry = malloc(sizeof(struct huge_alloc));
+	entry->fd = hugepage_fd;
+	strncpy(entry->path, hugepage_path, sizeof(entry->path) - 1);
+	entry->addr = buf;
+	entry->len = nbytes;
+	SLIST_INSERT_HEAD(&huge_alloc_head, entry, link);
+
+	return buf;
+}
+
+void
+xnvme_be_linux_buf_free(const struct xnvme_dev *XNVME_UNUSED(dev), void *buf)
+{
+	struct huge_alloc *entry;
+
+	// Iterate through linked-list to find the entry from *buf
+	entry = SLIST_FIRST(&huge_alloc_head);
+	while (entry != NULL) {
+		if (entry->addr == buf) {
+			break;
+		}
+		entry = SLIST_NEXT(entry, link);
+	}
+
+	if (!entry) {
+		XNVME_DEBUG("huge_alloc_table_reset: Entry %p not found!", buf);
+	}
+
+	// Clean up all the files, buffers etc. for the hugepage
+	munmap(buf, entry->len);
+	close(entry->fd);
+	remove(entry->path);
+	SLIST_REMOVE(&huge_alloc_head, entry, huge_alloc, link);
+	free(entry);
+}
+
+#endif
+
+struct xnvme_be_mem g_xnvme_be_linux_mem = {
+	.id = "hugepage",
```

Comment:
For the alloc/free functions, currently named:
```
xnvme_be_linux_buf_alloc()
xnvme_be_linux_buf_free()
```
These should be named something signifying that they are for hugepages. The backend interface is not that consistent here, but you have the option to do a object-local name, since the allocator is self-contained, so they could be named:

```
_hugepage_buf_alloc()
_hugepage_buf_free()
```
or using the full prefix:
```
xnvme_be_linux_mem_hugepage_buf_alloc()
xnvme_be_linux_mem_hugepage_buf_free()
```

Reply:
Picked the full name